File: MiniDOMBase/DMD/Luoghi/Regioni/CRegione.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases
Imports minidom.Anagrafica

Partial Public Class Anagrafica

    <Serializable> _
    Public Class CRegione
        Inherits Luogo

        Private m_Sigla As String
        Private m_NomeAbitanti As String
        Private m_Nazione As String
        Private m_NumeroAbitanti As Integer

        Public Sub New()
            Me.m_Sigla = ""
            Me.m_Nazione = ""
            Me.m_NomeAbitanti = ""
            Me.m_NumeroAbitanti = 0
        End Sub

        Public Overrides Function GetModule() As CModule
            Return Luoghi.Regioni.Module
        End Function

        Public Property NumeroAbitanti As Integer
            Get
                Return Me.m_NumeroAbitanti
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.m_NumeroAbitanti
                If (oldValue = value) Then Exit Property
                Me.m_NumeroAbitanti = value
                Me.DoChanged("NumeroAbitanti", value, oldValue)
            End Set
        End Property

        Public Property NomeAbitanti As String
            Get
                Return Me.m_NomeAbitanti
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_NomeAbitanti
                If (oldValue = value) Then Exit Property
                Me.m_NomeAbitanti = value
                Me.DoChanged("NomeAbitanti", value, oldValue)
            End Set
        End Property

        Public Property Sigla As String
            Get
                Return Me.m_Sigla
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_Sigla
                If (oldValue = value) Then Exit Property
                Me.m_Sigla = value
                Me.DoChanged("Sigla", value, oldValue)
            End Set
        End Property

        Public Property Nazione As String
            Get
                Return Me.m_Nazione
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_Nazione
                If (oldValue = value) Then Exit Property
                Me.m_Nazione = value
                Me.DoChanged("Regione", value, oldValue)
            End Set
        End Property

        Protected Friend Overrides Function GetConnection() As Databases.CDBConnection
            Return Anagrafica.Luoghi.Database
        End Function

        Public Overrides Function GetTableName() As String
            Return "tbl_Luoghi_Regioni"
        End Function

        Protected Overrides Function LoadFromRecordset(ByVal reader As DBReader) As Boolean
            Me.m_NumeroAbitanti = reader.Read("NumeroAbitanti", Me.m_NumeroAbitanti)
            Me.m_Sigla = reader.Read("Sigla", Me.m_Sigla)
            Me.m_Nazione = reader.Read("Nazione", Me.m_Nazione)
            Me.m_NomeAbitanti = reader.Read("NomeAbitanti", Me.m_NomeAbitanti)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(ByVal writer As DBWriter) As Boolean
            writer.Write("NumeroAbitanti", Me.m_NumeroAbitanti)
            writer.Write("Sigla", Me.m_Sigla)
            writer.Write("Nazione", Me.m_NomeAbitanti)
            writer.Write("NomeAbitanti", Me.m_NomeAbitanti)
            Return MyBase.SaveToRecordset(writer)
        End Function

        Public Overrides Function ToString() As String
            Return Me.Nome
        End Function

        Protected Overrides Sub XMLSerialize(ByVal writer As minidom.XML.XMLWriter)
            writer.WriteAttribute("NumeroAbitanti", Me.m_NumeroAbitanti)
            writer.WriteAttribute("Sigla", Me.m_Sigla)
            writer.WriteAttribute("Nazione", Me.m_Nazione)
            writer.WriteAttribute("NomeAbitanti", Me.m_NomeAbitanti)
            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(ByVal fieldName As String, ByVal fieldValue As Object)
            Select Case (fieldName)
                Case "NumeroAbitanti" : Me.m_NumeroAbitanti = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "Sigla" : Me.m_Sigla = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Nazione" : Me.m_Nazione = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "NomeAbitanti" : Me.m_NomeAbitanti = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select
        End Sub

        Protected Overrides Sub OnAfterSave(e As SystemEvent)
            MyBase.OnAfterSave(e)
            Luoghi.Regioni.UpdateCached(Me)
        End Sub
    End Class

End Class

File: MiniDOMBase/DMD/Luoghi/Regioni/CRegione.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;
using System.Collections.Generic;

namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Rappresenta una regione
        /// </summary>
        [Serializable]
        public class CRegione 
            : Luogo
        {
            private string m_Sigla;
            private string m_Nazione;
            
            /// <summary>
            /// Costruttore
            /// </summary>
            public CRegione()
            {
                m_Sigla = "";
                m_Nazione = "";
            }

            /// <summary>
            /// Repository
            /// </summary>
            /// <returns></returns>
            public override CModulesClass GetModule()
            {
                return minidom.Anagrafica.Luoghi.Regioni; //.Module;
            }

              
            /// <summary>
            /// Sigla
            /// </summary>
            public string Sigla
            {
                get
                {
                    return m_Sigla;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_Sigla;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Sigla = value;
                    DoChanged("Sigla", value, oldValue);
                }
            }

            /// <summary>
            /// Nazione
            /// </summary>
            public string Nazione
            {
                get
                {
                    return m_Nazione;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_Nazione;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Nazione = value;
                    DoChanged("Regione", value, oldValue);
                }
            }

            //protected internal override Databases.CDBConnection GetConnection()
            //{
            //    return Luoghi.Database;
            //}

            /// <summary>
            /// Discriminator
            /// </summary>
            /// <returns></returns>
            public override string GetTableName()
            {
                return "tbl_Luoghi_Regioni";
            }

            /// <summary>
            /// Carica dal db
            /// </summary>
            /// <param name="reader"></param>
            /// <returns></returns>
            protected override bool LoadFromRecordset(DBReader reader)
            {
                m_Sigla = reader.Read("Sigla", m_Sigla);
                m_Nazione = reader.Read("Nazione", m_Nazione);
                return base.LoadFromRecordset(reader);
            }

            /// <summary>
            /// Salva nel db
            /// </summary>
            /// <param name="writer"></param>
            /// <returns></returns>
            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("Sigla", this.m_Sigla);
                writer.Write("Nazione", this.m_Nazione);
                return base.SaveToRecordset(writer);
            }

            /// <summary>
            /// Prepara lo schema
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaFields(DBTable table)
            {
                base.PrepareDBSchemaFields(table);

                var c = table.Fields.Ensure("Sigla", typeof(string), 255);
                c = table.Fields.Ensure("Nazione", typeof(string), 255);
            }

            /// <summary>
            /// Prepara i vincoli
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaConstraints(DBTable table)
            {
                base.PrepareDBSchemaConstraints(table);

                var c = table.Constraints.Ensure("idxNazione", new string[] { "Sigla", "Nazione" }, DBFieldConstraintFlags.None);
            }

            /// <summary>
            /// Restituisce una stringa che rappresenta l'oggetto
            /// </summary>
            /// <returns></returns>
            public override string ToString()
            {
                return Nome;
            }

            /// <summary>
            /// Restituisce il codice hash dell'oggetto
            /// </summary>
            /// <returns></returns>
            public override int GetHashCode()
            {
                return base.GetHashCode();
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Sigla", m_Sigla);
                writer.WriteAttribute("Nazione", m_Nazione);
                base.XMLSerialize(writer);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "Sigla":
                        {
                            m_Sigla = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "Nazione":
                        {
                            m_Nazione = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                   

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public sealed override bool Equals(Luogo obj)
            {
                return (obj is CRegione) && this.Equals((CRegione)obj);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public virtual bool Equals(CRegione obj)
            {
                return base.Equals(obj)
                    && DMD.Strings.EQ(this.m_Sigla, obj.m_Sigla)
                    && DMD.Strings.EQ(this.m_Nazione, obj.m_Nazione)
                    ;
            }

        }
    }
}